server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/DeleteMailHelper.java

```diff
@@ -0,0 +1,86 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.rabbitmq.view.cassandra;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.queue.rabbitmq.MailQueueName;
+import org.apache.james.queue.rabbitmq.view.cassandra.model.EnqueuedMail;
+import org.apache.james.queue.rabbitmq.view.cassandra.model.MailKey;
+import org.apache.mailet.Mail;
+
+class DeleteMailHelper {
+
+    private final DeletedMailsDAO deletedMailsDao;
+    private final BrowseStartDAO browseStartDao;
+    private final BrowseHelper browseHelper;
+    private final CassandraMailQueueViewConfiguration configuration;
+    private final Random random;
+
+    @Inject
+    DeleteMailHelper(DeletedMailsDAO deletedMailsDao,
+                     BrowseStartDAO browseStartDao,
+                     BrowseHelper browseHelper, CassandraMailQueueViewConfiguration configuration) {
+        this.deletedMailsDao = deletedMailsDao;
+        this.browseStartDao = browseStartDao;
+        this.browseHelper = browseHelper;
+
+        this.configuration = configuration;
+        this.random = new Random();
+    }
+
+    CompletableFuture<Void> updateDeleteTable(Mail mail, MailQueueName mailQueueName) {
+        return deletedMailsDao
+            .markAsDeleted(mailQueueName, MailKey.fromMail(mail))
+            .thenRunAsync(() -> updateBrowseStart(mailQueueName));
+    }
+
+    private void updateBrowseStart(MailQueueName mailQueueName) {
+        if (shouldBrowseStart()) {
+            findNewBrowseStart(mailQueueName)
+                .thenCompose(newBrowseStart -> setNewBrowseStart(mailQueueName, newBrowseStart))
+                .join();
+        }
+    }
+
+    private CompletableFuture<Optional<Instant>> findNewBrowseStart(MailQueueName mailQueueName) {
+        return browseHelper.browseReferences(mailQueueName)
+            .map(EnqueuedMail::getTimeRangeStart)
+            .completableFuture()
+            .thenApply(Stream::findFirst);
+    }
+
+    private CompletableFuture<Void> setNewBrowseStart(MailQueueName mailQueueName, Optional<Instant> newBrowseStart) {
+        return newBrowseStart.map(value ->
+            browseStartDao.updateBrowseStart(mailQueueName, value))
+            .orElse(CompletableFuture.completedFuture(null));
+    }
+
+    private boolean shouldBrowseStart() {
+        int threshold = configuration.getUpdateFirstEnqueuedPace();
+        return Math.abs(random.nextInt()) % threshold == 0;
```

Comment:
hope we'll be lucky :trollface:

Reply:
Well, "large number law" is here for us an randomness has excellent distributed and concurrent properties.

That, is an elegant mathematical solution to avoid doing an expensive computation all the time.

If I did not convinced you yet think read repair in Cassandra, and the likes.